src/Mod/Import/App/dxf/dxf.cpp

```diff
@@ -3271,7 +3280,118 @@ bool CDxfRead::ReadLayer()
     return false;
 }
 
-void CDxfRead::DoRead(const bool ignore_errors /* = false */ )
+bool CDxfRead::ReadVersion()
+{
+    static const std::vector<std::string> VersionNames = {
+        // This table is indexed by eDXFVersion_t - (ROlder+1)
+        "AC1006",
+        "AC1009",
+        "AC1012",
+        "AC1014",
+        "AC1015",
+        "AC1018",
+        "AC1021",
+        "AC1024",
+        "AC1027",
+        "AC1032"};
+
+    assert(VersionNames.size() == RNewer - ROlder - 1);
+    get_line();
+    get_line();
+    std::vector<std::string>::const_iterator first = VersionNames.cbegin();
+    std::vector<std::string>::const_iterator last = VersionNames.cend();
+    std::vector<std::string>::const_iterator found = std::lower_bound(first, last, m_str);
+    if (found == last)
+        m_version = RNewer;
+    else if (*found == m_str)
+        m_version = (eDXFVersion_t)(std::distance(first, found) + (ROlder + 1));
+    else if (found == first)
+        m_version = ROlder;
+    else
+        m_version = RUnknown;
+
+    return ResolveEncoding();
+}
+
+bool CDxfRead::ReadDWGCodePage()
+{
+    get_line();
+    get_line();
+    assert(m_CodePage == NULL); // If not, we have found two DWGCODEPAGE variables or DoRead was called twice on the same CDxfRead object.
+    m_CodePage = new std::string(m_str);
+
+    return ResolveEncoding();
+}
+
+bool CDxfRead::ResolveEncoding()
+{
+    if (m_encoding != NULL) {
+        delete m_encoding;
+        m_encoding = NULL;
+    }
+    if (m_version >= R2007) {   // Note this does not include RUnknown, but does include RLater
+        m_encoding = new std::string("utf_8");
+        stringToUTF8 = &CDxfRead::UTF8ToUTF8;
+    }
+    else if (m_CodePage == NULL) {
+        // cp1252
+        m_encoding = new std::string("cp1252");
+        stringToUTF8 = &CDxfRead::GeneralToUTF8;
+    }
+    else {
+        // Names may be of the form "ansi_1252" which we map to "cp1252" but we don't map "ansi_x3xxxx" which means "ascii"
+        if (strncmp(m_CodePage->c_str(), "ansi_", 5) == 0
+            && strncmp(m_CodePage->c_str(), "ansi_x3", 7) != 0) {
+            std::string* p = new std::string(*m_CodePage);
+            p->replace(0, 5, "cp");
+            m_encoding = p;
+        }
+        else
+            m_encoding = new std::string(*m_CodePage);
+        // At this point we want to recognize synonyms for "utf_8" and use the custom decoder function.
+        // This is because this is one of the common cases and our decoder function is a fast no-op.
+        // We don't actually use the decoder function we get from PyCodec_Decoder because to call it we have to convert the (char *) text into
+        // a 'bytes' object first so we can pass it to the function using PyObject_Callxxx(), getting the PYObject containing the
+        // Python string, which we then decode back to UTF-8. It is simpler to call PyUnicode_DecodeXxxx which takes a (const char *)
+        // and is just a direct c++ callable.
```

Comment:
Before using any Python functions you must lock the GIL. Do this by adding the line
`Base::PyGILStateLocker lock;`

Reply:
How much performance penalty does this have? Say, compared to the cost of the call(s) that it is protecting?

Just asking because, if it is very expensive (either in processor or real time) I might consider some other encoding conversion method. My assumption is that this asserts some sort of lock, so the only impact would be if garbage collection is in progress, there would be a real-time delay waiting for it to complete?

Reply:
Locking code added

Reply:
No, it's not very expensive. If the GIL is not locked and the used Python function requires it then Python will call abort() and crashes the application.